### PR TITLE
Deprecate in-tree `rand`, `std::rand` and `#[derive(Rand)]`.

### PR DESCRIPTION
**Pull request by huonw**
Use [`rand`](https://crates.io/crates/rand) and [`derive_rand`](https://crates.io/crates/derive_rand) from crates.io.

[breaking-change]

As described in the RFC: https://github.com/rust-lang/rfcs/pull/722#issuecomment-71779521

### CODE REVIEW

File: src/libsyntax/ext/deriving/rand.rs [hunk 1]
```diff
@@ -24,6 +24,9 @@ pub fn expand_deriving_rand<F>(cx: &mut ExtCtxt,
                                push: F) where
     F: FnOnce(P<Item>),
 {
+    cx.span_warn(span,
+                 "`derive(Rand)` is deprecated in favour of `derive_rand` on crates.io");
```
Comment (alexcrichton) [inline src/libsyntax/ext/deriving/rand.rs]: Has this crate not been pushed to `rust-lang/rand` just yet? I see it on crates.io but not in git

Comment (huonw) [inline src/libsyntax/ext/deriving/rand.rs]: Whoops, pushed now.

Comment (huonw) [inline src/libsyntax/ext/deriving/rand.rs]: (I'm wondering if `rand_macros` may be a better name...)

Comment (alexcrichton) [inline src/libsyntax/ext/deriving/rand.rs]: Hm yeah, I might go with `rand_macros` instead, but it's not super critical.

**Pull request merged**